File: c-sharp/Models/ScalarResults.cs
using System.Collections;
using System.Collections.Generic;

namespace HmxLabs.TechTest.Models
{
    public class ScalarResults : IScalarResultReceiver, IEnumerable<ScalarResult>
    {
        public ScalarResult? this[string tradeId_]
        {
            get
            {
                if (!ContainsTrade(tradeId_))
                {
                    return null;
                }

                double? priceResult = null;
                string? error = null;
                if (_results.ContainsKey(tradeId_))
                {
                    priceResult = _results[tradeId_];
                }
                if (_errors.ContainsKey(tradeId_))
                {
                    error = _errors[tradeId_];
                }

                return new ScalarResult(tradeId_, priceResult, error);
            }
        }

        public bool ContainsTrade(string tradeId_)
        {
            if (_results.ContainsKey(tradeId_) || _errors.ContainsKey(tradeId_))
            {
                return true;
            }

            return false;
        }

        public void AddResult(string tradeId_, double result_)
        {
            _results.Add(tradeId_, result_);
        }

        public void AddError(string tradeId_, string? error_)
        {
            _errors.Add(tradeId_, error_);
        }

        public IEnumerator<ScalarResult> GetEnumerator()
        {
            // Enumerate results first
            foreach (var (tradeId, value) in _results)
            {
                string? error = null;
                if (_errors.TryGetValue(tradeId, out var e))
                    error = e;

                yield return new ScalarResult(tradeId, value, error);
            }

            // Then enumerate errors that don't have a result
            foreach (var (tradeId, value) in _errors)
            {
                if (_results.ContainsKey(tradeId))
                    continue;

                yield return new ScalarResult(tradeId, null, value);
            }
        }

        IEnumerator IEnumerable.GetEnumerator()
        {
            return GetEnumerator();
        }

        private readonly Dictionary<string, double> _results = new Dictionary<string, double>();
        private readonly Dictionary<string, string?> _errors = new Dictionary<string, string?>();
    }
}

File: cpp/Models/ScalarResults.cpp
#include "ScalarResults.h"

#include <map>
#include <stdexcept>

ScalarResults::~ScalarResults() = default;

std::optional<ScalarResult> ScalarResults::operator[](const std::string& tradeId) const {
    if (!containsTrade(tradeId)) {
        return std::nullopt;
    }

    std::optional<double> priceResult = std::nullopt;
    std::optional<std::string> error = std::nullopt;

    if (const auto resultIt = results_.find(tradeId); resultIt != results_.end()) {
        priceResult = resultIt->second;
    }

    if (const auto errorIt = errors_.find(tradeId); errorIt != errors_.end()) {
        error = errorIt->second;
    }

    return ScalarResult(tradeId, priceResult, error);
}

bool ScalarResults::containsTrade(const std::string& tradeId) const {
    return results_.find(tradeId) != results_.end() || errors_.find(tradeId) != errors_.end();
}

void ScalarResults::addResult(const std::string& tradeId, double result) {
    results_[tradeId] = result;
}

void ScalarResults::addError(const std::string& tradeId, const std::string& error) {
    errors_[tradeId] = error;
}

ScalarResults::Iterator::Iterator(
    const ScalarResults* parent,
    std::shared_ptr<const std::vector<ScalarResult>> snapshot,
    const std::size_t index)
    : parent_(parent), snapshot_(std::move(snapshot)), index_(index) {
}

ScalarResults::Iterator& ScalarResults::Iterator::operator++() {
    if (snapshot_ != nullptr && index_ < snapshot_->size()) {
        ++index_;
    }
    return *this;
}

ScalarResults::Iterator::reference ScalarResults::Iterator::operator*() const {
    if (parent_ == nullptr || snapshot_ == nullptr || index_ >= snapshot_->size()) {
        throw std::out_of_range("Iterator cannot be dereferenced");
    }

    return (*snapshot_)[index_];
}

ScalarResults::Iterator::pointer ScalarResults::Iterator::operator->() const {
    if (parent_ == nullptr || snapshot_ == nullptr || index_ >= snapshot_->size()) {
        throw std::out_of_range("Iterator cannot be dereferenced");
    }

    return &(*snapshot_)[index_];
}

bool ScalarResults::Iterator::operator==(const Iterator& other) const {
    const bool thisIsEnd = snapshot_ == nullptr || index_ >= snapshot_->size();

    if (const bool otherIsEnd = other.snapshot_ == nullptr || other.index_ >= other.snapshot_->size(); thisIsEnd || otherIsEnd) {
        return thisIsEnd == otherIsEnd;
    }

    return parent_ == other.parent_ && index_ == other.index_;
}

bool ScalarResults::Iterator::operator!=(const Iterator& other) const {
    return !(*this == other);
}

std::shared_ptr<const std::vector<ScalarResult>> ScalarResults::buildSnapshot() const {
    std::map<std::string, std::pair<std::optional<double>, std::optional<std::string>>> mergedResults;

    for (const auto& [fst, snd] : results_) {
        mergedResults[fst].first = snd;
    }

    for (const auto& [fst, snd] : errors_) {
        mergedResults[fst].second = snd;
    }

    std::vector<ScalarResult> snapshot;
    snapshot.reserve(mergedResults.size());

    for (const auto& [tradeId, resultData] : mergedResults) {
        snapshot.emplace_back(tradeId, resultData.first, resultData.second);
    }

    return std::make_shared<const std::vector<ScalarResult>>(std::move(snapshot));
}

ScalarResults::Iterator ScalarResults::begin() const {
    // The iterator walks an immutable snapshot so copied iterators remain valid
    // and comparisons are based on logical position rather than snapshot identity.
    return Iterator(this, buildSnapshot(), 0);
}

ScalarResults::Iterator ScalarResults::end()
{
    return Iterator();
}
